.github/workflows/check_smoke_tests.yml

```diff
@@ -7,132 +7,182 @@ on:
     branches: [main]
 
 jobs:
+  service-set-up:
+    runs-on: ubuntu-latest
+
+    strategy:
+      fail-fast: true
+      matrix:
+        database: [postgres:17, mysql:8, mssql/server:2022-latest]
+
+    steps:
+      - name: Check out repository
+        uses: actions/checkout@v4
+
+      - name: Convert repository name to lowercase
+        run: echo "REPO=$(echo '${{ github.repository }}' | tr '[:upper:]' '[:lower:]')" >> $GITHUB_ENV
+
+      - name: Log in to GitHub Container Registry
+        uses: docker/login-action@v3
+        with:
+          registry: ghcr.io
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Pull and push db images to GHCR if necessary
+        run: |
+          # If mssql, use a slightly different strategy
+          if [[ "${{ matrix.database }}" == "mssql/server:2022-latest" ]]; then
+            TARGET_IMAGE="ghcr.io/${{ env.REPO }}/mssql:2022-latest"
+            SOURCE_IMAGE="mcr.microsoft.com/mssql/server:2022-latest"
+          else
+            TARGET_IMAGE="ghcr.io/${{ env.REPO }}/${{ matrix.database }}"
+            SOURCE_IMAGE="${{ matrix.database }}"
+          fi
+
+          # Check if the image exists in GHCR
+          if ! docker pull "$TARGET_IMAGE" >/dev/null; then
+            echo "Image not found in GHCR, pulling from source: $SOURCE_IMAGE..."
+            if docker pull "$SOURCE_IMAGE"; then
```

Comment:
Something to consider, we are also using database images in the check_unit_tests workflow.  I think with this strategy we'd have to copy these jobs there as well.  Is that right?